haskell-src/Concordium/Genesis/Data.hs

```diff
@@ -68,3 +69,11 @@ putVersionedGenesisData = case protocolVersion @pv of
 genesisBlockHash :: forall pv. IsProtocolVersion pv => GenesisData pv -> BlockHash
 genesisBlockHash = case protocolVersion @pv of
     SP1 -> P1.genesisBlockHash . unGDP1
+
+-- |A dependent pair of a protocol version and genesis data.
+data PVGenesisData = forall pv. IsProtocolVersion pv => PVGenesisData (GenesisData pv)
```

Comment:
Isn't this just an existential type, not a dependent pair?

Also here you use the existential syntax but for the SomeProtocolVersion you use the GADT syntax.

Reply:
I think of this as being essentially the sum type Σₚ GenesisData(p). Because of the `IsProtocolVersion` constraint, the protocol version is available at runtime.